cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,244 @@
+    this.schema = schema;
+    this.recordWriter = recordWriter;
+    this.aggregationWindowSeconds = Math.max(1, aggregationWindowSeconds);
+  }
+
+  @Override
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    cachemap.clear();
+    if (!input.hasNext()) {
+      return input;
+    }
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    // MetricValues will be aggregated across windows of size "aggregationWindowSeconds".
+    // We align the current time to the end of the last widow. This is done since in most cases,
+    // it is expected that all the MetricValues were emitted very recently. The offset below ensures
+    // the last window completely lies in the past and maximizes the chance that all metrics lie within it.
+    long offset = (currentTimestamp % aggregationWindowSeconds) + 1;
+
+    input.forEachRemaining((message) -> {
+      // Decode the metrics.
+      // Keep track of the last message ID as TopicRelayers need it to fetch messages in the next iteration.
+      messageId[0] = message.getId();
+      decodingInputStream.reset(message.getPayload());
+      MetricValues values;
+      try {
+        values = reader.read(decoder, schema);
+      } catch (IOException e) {
+        LOG.warn("Failed to decode metric message.", e);
+        return;
+      }
+
+      // Aggregate the metrics.
+      long timestamp = values.getTimestamp();
+      MetricsAggregator metricValuesAggregator = aggregated.computeIfAbsent(
+        new MetricsMapKey(values.getTags(), getWindowNumber(values.getTimestamp(), offset)),
+        (t) -> new MetricsAggregator(timestamp));
+      for (MetricValue metric : values.getMetrics()) {
+        metricValuesAggregator.addMetric(metric, timestamp);
+      }
+    });
+
+    List<Message> res = aggregated
+      .entrySet()
+      .stream()
+      .map((entry) -> entry.getValue().getAggregateMetricValues(entry.getKey().getTags()))
+      .map((metricValues) -> {
+        // Encode the metric value into a message.
+        encodeOutputStream.reset();
+        try {
+          recordWriter.encode(metricValues, encoder);
+        } catch (IOException e) {
+          throw new RuntimeException("Failed to encode metric to message payload.", e);
+        }
+        byte[] payload = encodeOutputStream.toByteArray();
+        return new Message() {
+          @Override
+          public String getId() {
+            // TopicRelayers keep track of the last message ID
+            // to fetch messages in the next iteration.
+            // Message IDs are generated by the messaging service. Return the last message ID before aggregation
+            // every time.
+            return messageId[0];
+          }
+
+          @Override
+          public byte[] getPayload() {
+            return payload;
+          }
+        };
+      })
+      .collect(Collectors.toList());
```

Comment:
You don't need to collect. You can get an `Iterator` from a `Stream` directly by call `Stream.iterator()`.

Reply:
Done, I was seeing compiler errors while returning an iterator directly. I was able to resolve them by casting the anonymous struct to `Message`.